owltorrent-network/src/main/java/edu/rice/owltorrent/network/TorrentRepositoryImpl.java

```diff
@@ -0,0 +1,39 @@
+package edu.rice.owltorrent.network;
+
+import edu.rice.owltorrent.common.entity.TwentyByteId;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2(topic = "general")
+public class TorrentRepositoryImpl implements TorrentRepository {
+  private ConcurrentHashMap<TwentyByteId, TorrentManager> infoHashToTorrentManager;
+
+  public TorrentRepositoryImpl() {
+    infoHashToTorrentManager = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void registerTorrentManager(TorrentManager manager) {
+    TwentyByteId infoHash = manager.getTorrent().getInfoHash();
+    /* This conditional should not be triggered in any case. */
+    if (infoHashToTorrentManager.containsKey(infoHash)) {
+      log.error("Torrent manager for torrent {} already exist.", manager.getTorrent().getName());
+      return;
+    }
+    this.infoHashToTorrentManager.put(infoHash, manager);
```

Comment:
This shouldn't matter much, but checking if contains or not and then put is not thread safe. You want to use something like putIfAbsent instead.